### PR TITLE
Test and refactor RdsResultStorage

### PR DESCRIPTION
**Pull request by julien-sebire**
To ease the schema change summoned by the SPanner benchmark, a refactoring of this file should be made to have the fewest change in the future.
All the methods to be refactored has been covered with unit testing prior to refactor.
Refactoring was done with factorizing code and using querybuilder as much as possible.

Better follow the different commits to see the successive changes.
The version bump is minimal since it's only testing and refactoring.

### CODE REVIEW

File: test/unit/model/RdsResultStorageTest.php [hunk 1]
```diff
@@ -83,72 +82,408 @@ public function testStoreRelatedDelivery()
...
+        $this->instance->storeRelatedDelivery($deliveryResultIdentifier2, $deliveryIdentifier2);
+
+        $this->assertEquals(2, $this->instance->countResultByDelivery([]));
+        $this->assertEquals(1, $this->instance->countResultByDelivery($deliveryIdentifier1));
+    }
+
+    /**
+     * @dataProvider resultByDeliveryToTest
+     *
+     * @param $ids
```
Comment (ekkinox) [inline test/unit/model/RdsResultStorageTest.php]: if you doc it, maybe you can specify the type ?

Comment (julien-sebire) [reply test/unit/model/RdsResultStorageTest.php]: Right. Done.

File: model/RdsResultStorage.php [hunk 2]
```diff
@@ -189,81 +136,48 @@ public function storeItemVariables(
...
-        
+
         $this->getPersistence()->insertMultiple(self::VARIABLES_TABLENAME, $dataToInsert);
     }
 
-    /*
-     * retrieve specific parameters from the resultserver to configure the storage
-     */
-    public function configure($callOptions = array())
+    public function storeRelatedTestTaker($deliveryResultIdentifier, $testTakerIdentifier)
```
Comment (peetya) [inline model/RdsResultStorage.php]: Why did you remove the docblock? (same about some other methods)

File: model/RdsResultStorage.php [hunk 3]
```diff
@@ -189,81 +136,48 @@ public function storeItemVariables(
...
-        $sql = 'SELECT COUNT(*) FROM ' . self::RESULTS_TABLENAME .
-            ' WHERE ' . self::RESULTS_TABLE_ID . ' = ?';
-        $params = array($deliveryResultIdentifier);
-        if ($this->getPersistence()->query($sql, $params)->fetchColumn() == 0) {
+        $qb = $this->getQueryBuilder()
+            ->select('COUNT(*)')
+            ->from(self::RESULTS_TABLENAME)
+            ->andWhere(self::RESULTS_TABLE_ID .' = :id')
+            ->setParameter('id', $deliveryResultIdentifier);
+        if ($qb->execute()->fetchColumn() == 0) {
```
Comment (peetya) [inline model/RdsResultStorage.php]: `===`

File: model/RdsResultStorage.php [hunk 4]
```diff
@@ -273,263 +187,260 @@ public function getVariables($callId)
...
-            $inQuery = implode(',', array_fill(0, count($delivery), '?'));
-            $sql .= self::DELIVERY_COLUMN . ' IN (' . $inQuery . ')';
-            $params = array_merge($params, $delivery);
+            $qb
+                ->andWhere(self::DELIVERY_COLUMN .' IN (:delivery)')
+                ->setParameter('delivery', $delivery, Connection::PARAM_STR_ARRAY);
         }
 
-        return $this->getPersistence()->query($sql, $params)->fetchColumn();
+        return $qb->execute()->fetchColumn();
```
Comment (peetya) [inline model/RdsResultStorage.php]: Is it safe to execute the `COUNT` even without `WHERE` clause? (when delivery array is empty)

File: model/RdsResultStorage.php (hunk 2 above)
Comment (julien-sebire) [reply model/RdsResultStorage.php]: Most of them were just replicating the interfaces' ones. I ran through all of them to check when it was necessary...

File: model/RdsResultStorage.php (hunk 3 above)
Comment (julien-sebire) [reply model/RdsResultStorage.php]: In fact, the result is returned as a string. I could test it this way or with `=== '0'`.
Maybe `(int) $qb->execute()->fetchColumn() === 0` is better.
What do you think?

File: model/RdsResultStorage.php (hunk 4 above)
Comment (julien-sebire) [reply model/RdsResultStorage.php]: What do you mean by safe? It then turns out to a count(*) on a table. It's covered in the test (line 131).
In fact, there is an impact on Spanner performance though, since Spanner doesn't store statistics on the tables (because they're sharded), and for big tables that would turn into long query execution. But this is on the side of optimisation.

File: model/RdsResultStorage.php (hunk 3 above)
Comment (peetya) [reply model/RdsResultStorage.php]: It's always more safe if you check the type as well. Just think about `null == 0 vs null === 0`, etc... Personally I would prefer the 2nd version.

Comment (julien-sebire) [reply model/RdsResultStorage.php]: Ok, fixed!

File: test/unit/model/RdsResultStorageTest.php [hunk 5]
```diff
@@ -83,72 +82,408 @@ public function testStoreRelatedDelivery()
...
+        $this->instance->storeRelatedDelivery($deliveryResultIdentifier1, $deliveryIdentifier1);
+        $this->instance->storeRelatedTestTaker($deliveryResultIdentifier2, $testTakerIdentifier2);
+        $this->instance->storeRelatedDelivery($deliveryResultIdentifier2, $deliveryIdentifier2);
+
+        $this->assertEquals(2, $this->instance->countResultByDelivery([]));
+        $this->assertTrue($this->instance->deleteResult($deliveryResultIdentifier1));
+        $this->assertEquals(1, $this->instance->countResultByDelivery([]));
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
```
Comment (mike-ionut-mihai-sandu-tao) [inline test/unit/model/RdsResultStorageTest.php]: I think this line should not be here

Comment (julien-sebire) [reply test/unit/model/RdsResultStorageTest.php]: Yeah right, I first tried to split between item and test saving, but this made too many changes for the PR to be easy to read so I forgot about it.
Removed.
